### ISSUE TITLE
LPS lives left countdown announcement

### ISSUE THREAD

**Issue by kai-li-wop** [PadWorld-Entertainment/worldofpadman#171]
Similar to the existing points and time limit, a countdown could be issued for the LPS game mode, in which the last 3 lives are indicated. This means that each player would have to be announced individually when he has only 3, 2 and 1 lives left. WoP has sound files for this since the PadMod, but they were never implemented to the code. I guess that this is a planned feature, but it has not yet been implemented. I am not sure how the code could look like but I think it could be located in cg_playerstate.c. Sound samples are already available:

- sound/feedback/countdown/3_lives.wav
- sound/feedback/countdown/2_lives.wav
- sound/feedback/countdown/1_life.wav

Help wanted how the if statement could look like to use the players livesleft information. I don't have a clue. ;)

```
if (cgs.gametype == GT_LPS) {
	if (player livesleft status == 3) {
		CG_AddBufferedSound(cgs.media.threeLivesLeftSound);
	} else if (player livesleft status == 2) {
		CG_AddBufferedSound(cgs.media.twoLivesLeftSound);
	} else if (player livesleft status == 1) {
		CG_AddBufferedSound(cgs.media.oneLifeLeftSound);
	}
}

```

**Comment by 3aTmE** [comment]
Just tried it on the recent codebase with a local game with 2 clients.

**TL;DR : Announcer works, starts one life too early, skips the third life completely.**

The announcer works. 

For some weird reason the countdown begins too early on **4** lives "You have **3** Lives left".

After getting fragged again the scoreboard jumps suddenly down to **2** (The third life gets ignored completely)  and the right announcer sound-file plays. "You have 2 lives left".

Getting fragged again also plays the right announcer sound-file "you have 1 life left"

When all lives are used up the right sound-file will be played "Better luck next time , maybe!"

**Comment by kai-li-wop** [comment]
Hmm, this is indeed a strange behavior. i just tested local game with 2 bots and everything works just fine. After losing 4th life announcer says 3 lives left, after losing next one, 2 lives left, after losing this, 1life left followed by "you lose" or "better luck next time" when losing last life.

This is the code which I adopted from fraglimit warnings. Not sure if I twisted something there. Looking at @mgerhardy. ;)
```
	//lifelimit warnings
	if (cgs.lpsStartLives > 0 && (cgs.gametype == GT_LPS) && !(cgs.lpsflags & LPSF_PPOINTLIMIT)) {
		livesLeft = (cg.snap->ps.stats[STAT_LIVESLEFT]);
		if (!(cg.lifelimitWarnings & 4) && livesLeft == 1) {
			cg.lifelimitWarnings |= 1 | 2 | 4;
			CG_AddBufferedSound(cgs.media.oneLifeSound);
		} else if (cgs.lpsStartLives > 2 && !(cg.lifelimitWarnings & 2) && livesLeft == 2) {
			cg.lifelimitWarnings |= 1 | 2;
			CG_AddBufferedSound(cgs.media.twoLivesSound);
		} else if (cgs.lpsStartLives > 3 && !(cg.lifelimitWarnings & 1) && livesLeft == 3) {
			cg.lifelimitWarnings |= 1;
			CG_AddBufferedSound(cgs.media.threeLivesSound);
		}
	}

```

**Comment by tuxaddict** [comment]
Tested on ` "version" is:"wop 1.7.0 dev linux-x86_64-debug^7"` Anteroom, 4 Bots, local game:

**1. Voice counting down:** 
This **worked fine** in my 3 tests as well. After loosing life no.4, the friendly announcer informed me: "_You have three lives left_" This worked till the bitter end with the known: "_Better luck next time, maybe_"

**2. Synchronization with scoreboard:**
I experienced a bit strange behavior as well: When loosing my live, during the correct  voice announcement, the **scoreboard still showed the lost life**, and not the new, current state indicated by the voice. To be clearer: It showed "4 Lives" while the voice correctly announced three left, so it shows the state prior to the voice announcement.

After the scoreboard disappeared, when i pressed Tab to check the Scoreboard again, everything was shown correct. So i would wildly guess, there is something in the order of the routines Scoreboard Voice that my be not fitting yet.
Maybe the quoted code block "interferes" with a "scoreboard is shown because you lost a life in LPS" or so routine.

best, me! (-:

**Comment by kai-li-wop** [comment]
Hi @tuxaddict, welcome! The thing is, that the announcer sounds are played in case of your own "death". The scoreboard is going to be updated after respawn (after losing a life). And there is the mismatch but it is not a bug imho. This is how the thing is programmed, same with point limit in FFA. The difference here is, that the point limit announcements are played if someone else just died. In LPS it is you that lost the life.

**Comment by mgerhardy** [comment]
what were your initial start lives while testing this feature?

btw. g_LPS_startlives has 10 as default in the code - but 8 in rotation.cfg - shouldn't this be unified?

**Comment by mgerhardy** [comment]
`g_LPS_startlives` is empty (as not yet set via game code) if you start the lps mode from the ui

**Comment by kai-li-wop** [comment]
With ea504badbaff99459d204409836210b6ec79ecc3 I updated the rotation.cfg and set 10 LPS lives. The start server UI sets `g_LPS_startlives` to 10 by default. I have tested the feature with 10, 5, 4, 3, 2, and 1 start lives set in the menu field and and could not find any problems with this setup. But in fact, I think it is possible to enter 0 into the field which would not make any sense. I need to test this further.

**Comment by kai-li-wop** [comment]
OK, with `g_LPS_startlives` set to 0 in the start server (create) menu you start as spectator waiting for next round probably forever. OK, I did not add some bots. ;)

**Comment by kai-li-wop** [comment]
Not directly connected to this issue but with b47967bb95cf57bf0a6afc2003095ea3d589f658 I tried to fix the missing default value in the start server (create) menu (UI). To test this you need to delete your wopconfig to start with default value of 10. After initial start the menu should show now "Lives Left: 10" instead of an empty entry field. Values below 1 should not be possible any more. You can still enter a value of 0 but it should be without effect.

So please test again:
- LPS announcer feature
- Start Server menu Start Lives entry field

**Comment by tuxaddict** [comment]
both tested, interesting results (-:
`"version" is:"wop 1.7.0 dev linux-x86_64-debug^7"`

### Start Server menu Start Lives entry field
| Test  | Result |
| ------------- | ------------- |
| Delete your wopconfig to start with default value  | true  |
| After initial start the menu should show now "Lives Left: 10"  | true  |
| You can still enter a value of 0 but it should be without effect | true  |
| Values below 1 should not be possible any more. | true |

### LPS announcer feature

- The **announcements work** as in the last test
- Only if the LPS lives are set to 1, the game starts with the announcment "You have only on live left"
- The finishing announcement **"Better luck next time" appears twice** only on 
**uneven numbers,** tested from 1-8, confirmed at 1,3,5,7
- The "**scoreboard is showning the prior to death value**" on announcmenets **still persists**, can be easily healed by pressing TAB, and, interestingly, also exists in my tests prior to the new voice announcements taking place

**Comment by 3aTmE** [comment]
I tested it once again, while the countdown now works reliable. 

Still there is some weirdness going on :

## LPS announcer feature testing : 

### **Starting LPS via the menu with lives set to 0 :**
- The game starts with 1 live. 
- Losing the live results in "better luck next time, maybe " being played twice.

### **Starting LPS via the menu with lives set to 1:**
- Same results!
- The game starts with 1 live. 
- Losing the live results in "better luck next time, maybe " being played twice.

###  **Starting LPS via the menu with 2 or more lives:**
- The game starts with the right amount of lives. 
- Losing the lives results in "better luck next time, maybe " being played **once**. The Round ends normally as expected.

### **Somehow bots (and possibly players) are able to get an score below 0 :**
![Image](https://user-images.githubusercontent.com/84514074/221410218-2fea1b2e-4e44-4140-9355-ce0608c3cdf7.jpg)
As for now i'm not sure what caused this behavior.

**Comment by kai-li-wop** [comment]
@3aTmE @tuxaddict: I committed a fix with further explanation (see 72c0f09dbfe6beecdcdb5c06135e54629f97abd5) which should prevent "better luck" played twice. This bug is not connected to the lives left countdown feature introduced with this enhancement ticket, but it should be fixed now. "you lose" should always only be played when you lose your last life, and "better luck" should always only be played at the end of the match when someone else wins the match. Waiting for your test and approval. Thx!

PS: I still don't know why players can have now -1 lives but it seems not to be connected to the "better luck" bug.

**Comment by tuxaddict** [comment]
> @3aTmE @tuxaddict: I committed a fix with further explanation (see [72c0f09](https://github.com/PadWorld-Entertainment/worldofpadman/commit/72c0f09dbfe6beecdcdb5c06135e54629f97abd5)) which should prevent "better luck" played twice. ... Waiting for your test and approval. Thx!

- After testing  several times, me can **approve** that the "better luck next time" phrase doesn't get triggered after loosing the last life anymore, and therefore doesn't appears twice at the end of a round.

Thanks for your time you putting into this beautiful game @kai-li-wop! (-:  

_btw: the brightness is working for me 1st time since i play wop, with 1.7.0 current branch! Great development done there, thumbs up!
As a side note, Interestingly the announcements of "N lives left" was missing completely on a few tests..._

**Comment by kai-li-wop** [comment]
One reason that a sound doesn't always play as expected could be that the sound buffer is full and the engine can't play all sounds in one channel at the same time. I sometimes have this problem when I collect Revival and use Floater at the same time. Floater then doesn't play any sound at all when I use it. I think it might be the same when playing a few announcer sounds at the same time. Some are just not audible.
